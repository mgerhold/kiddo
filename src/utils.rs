use std::ffi::OsStr;
use std::path::Path;

use bumpalo::Bump;

use crate::parser::errors::ParserError;
use crate::token::Token;

pub(crate) trait AllocPath {
    fn alloc_path(&self, path: impl AsRef<Path>) -> &Path;
}

impl AllocPath for Bump {
    fn alloc_path(&self, path: impl AsRef<Path>) -> &Path {
        Path::new(unsafe {
            OsStr::from_os_str_bytes_unchecked(
                self.alloc_slice_copy(path.as_ref().as_os_str().as_os_str_bytes()),
            )
        })
    }
}

pub(crate) fn parse_unsigned_int<'a, T: num_traits::Unsigned>(
    token: &'a Token<'a>,
) -> Result<T, ParserError<'a>> {
    let lexeme = token.lexeme();
    if let Some(lexeme) = lexeme.strip_prefix("0x") {
        T::from_str_radix(lexeme, 16).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else if let Some(lexeme) = lexeme.strip_prefix("0o") {
        T::from_str_radix(lexeme, 8).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else if let Some(lexeme) = lexeme.strip_prefix("0b") {
        T::from_str_radix(lexeme, 2).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else {
        T::from_str_radix(lexeme, 10).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    }
}
