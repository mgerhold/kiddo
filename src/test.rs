use utils::setup_test;
use utils::TestFile;

mod utils;

macro_rules! function_name {
    () => {{
        fn f() {}
        fn type_name_of<T>(_: T) -> &'static str {
            std::any::type_name::<T>()
        }
        let name = type_name_of(f);
        name.strip_suffix("::f")
            .unwrap()
            .split("::")
            .last()
            .unwrap()
    }};
}

#[test]
fn empty_source_file_compiles_successfully() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: "",
            },
            &[],
        )
        .compiler_output
    );
}

#[test]
fn importing_symbol_with_the_same_name_as_local_definition_fails() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"from main import f;

export function f() { }
"#,
            },
            &[],
        )
        .compiler_output
    );
}

#[test]
fn importing_symbol_from_other_module() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"from utils import f;
"#,
            },
            &[TestFile {
                filename: "utils.bs",
                source: r#"export function f() { }
"#
            }],
        )
        .compiler_output
    );
}

#[test]
fn global_variables() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"let global0 = 42;
let global1: U32 = 42;
let mutable global2 = 42;
let mutable global3: U32 = 42;
let const global4 = 42;
let const global5: U32 = 42;

export let global6 = 42;
export let global7: U32 = 42;
export let mutable global8 = 42;
export let mutable global9: U32 = 42;
export let const global10 = 42;
export let const global11: U32 = 42;
"#
            },
            &[]
        )
        .compiler_output
    );
}

#[test]
fn local_variables() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"function f() {
    let a = 10;
    let const b = 20;
    let mutable c = 30;
    let d: U32 = 40;
    let const e: U32 = 50;
    let mutable f: U32 = 60;
}
"#
            },
            &[]
        )
        .compiler_output
    );
}
