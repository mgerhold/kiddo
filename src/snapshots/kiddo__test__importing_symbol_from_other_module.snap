---
source: src/test.rs
expression: "setup_test(function!(),\n        TestFile { filename: \"main.bs\", source: r#\"from utils import f;\n\"# },\n        &[TestFile {\n                        filename: \"utils.bs\",\n                        source: r#\"export function f() { }\n\"#,\n                    }]).compiler_output"
---
[
    ModuleWithConnectedImports {
        filename: "main.bs",
        module: Module {
            imports: [
                FromImport {
                    where_: QualifiedNonTypeName(
                        Relative {
                            tokens: [
                                Token {
                                    source_location: "utils" (main.bs:1:6, byte_offset=5),
                                    type_: LowercaseIdentifier,
                                },
                            ],
                        },
                    ),
                    symbol: NonTypeIdentifier(
                        NonTypeIdentifier(
                            Token {
                                source_location: "f" (main.bs:1:19, byte_offset=18),
                                type_: LowercaseIdentifier,
                            },
                        ),
                    ),
                },
            ],
            definitions: [],
        },
        imports: [
            ConnectedImport {
                import: FromImport {
                    where_: QualifiedNonTypeName(
                        Relative {
                            tokens: [
                                Token {
                                    source_location: "utils" (main.bs:1:6, byte_offset=5),
                                    type_: LowercaseIdentifier,
                                },
                            ],
                        },
                    ),
                    symbol: NonTypeIdentifier(
                        NonTypeIdentifier(
                            Token {
                                source_location: "f" (main.bs:1:19, byte_offset=18),
                                type_: LowercaseIdentifier,
                            },
                        ),
                    ),
                },
                definition: Function(
                    FunctionDefinition {
                        is_exported: true,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "f" (utils.bs:1:17, byte_offset=16),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        parameters: [],
                        return_type: None,
                        body: Block {
                            statements: [],
                        },
                    },
                ),
            },
        ],
    },
    ModuleWithConnectedImports {
        filename: "utils.bs",
        module: Module {
            imports: [],
            definitions: [
                Function(
                    FunctionDefinition {
                        is_exported: true,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "f" (utils.bs:1:17, byte_offset=16),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        parameters: [],
                        return_type: None,
                        body: Block {
                            statements: [],
                        },
                    },
                ),
            ],
        },
        imports: [],
    },
]
